### DISCUSSION TITLE
Enable virtual threads preview support in Jetty

### DISCUSSION THREAD

**Discussion by olivierchedru** [dropwizard/dropwizard#6741] [General]
Jetty supports virtual threads preview: https://www.eclipse.org/jetty/documentation/jetty-11/operations-guide/index.html#og-server-threadpool-virtual
How to enable virtual threads preview in Dropwizard?
Thanks!

**Comment by zUniQueX** [comment]
Hi @olivierchedru. We don't provide support for virtual threads yet.

We configure the server in Java. So registering the module won't work. As far as I can see, the module you linked sets an other `ThreadPool` for the Jetty `Server`. Dropwizard already uses a custom `ThreadPool` to get metrics for Jetty.

To enable virtual thread support you currently would need to override the method `AbstractServerFactory#createThreadPool(MetricRegistry)`. There you can provide a custom `ThreadPool` which utilises virtual threads.

**Comment by olivierchedru** [comment]
Thank you. I managed to make it work this way:

- Create a resource `META-INF/services/io.dropwizard.core.server.ServerFactory` containing the class name of my custom server factory: `com.foo.MyServerFactory`.
- Define `MyServerFactory` with `virtual` type (example written in Kotlin):
```kotlin
/**
 * Server factory using virtual thread pools
 * @see <a href="https://www.eclipse.org/jetty/documentation/jetty-11/operations-guide/index.html#og-server-threadpool-virtual">Jetty virtual threads preview</a>
 */
@JsonTypeName("virtual")
class MyServerFactory : DefaultServerFactory() {

    override fun createThreadPool(metricRegistry: MetricRegistry): ThreadPool {
        val threadPool = super.createThreadPool(metricRegistry)
        if (threadPool is QueuedThreadPool) {
            threadPool.isUseVirtualThreads = true
        }
        return threadPool
    }
}
```
- Use `virtual` server type in `configuration.yaml`:
```yaml
server:
  type: virtual
  shutdownGracePeriod: 60s
  applicationConnectors:
  - type: h2c
    port: 8080
...
```
